velox/functions/sparksql/aggregates/BloomFilterAggAggregate.cpp

```diff
@@ -0,0 +1,313 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/sparksql/aggregates/BloomFilterAggAggregate.h"
+
+#include "velox/common/base/BloomFilter.h"
+#include "velox/exec/Aggregate.h"
+#include "velox/expression/FunctionSignature.h"
+#include "velox/vector/FlatVector.h"
+
+namespace facebook::velox::functions::sparksql::aggregates {
+
+namespace {
+
+struct BloomFilterAccumulator {
+  explicit BloomFilterAccumulator(HashStringAllocator* allocator)
+      : bloomFilter{StlAllocator<uint64_t>(allocator)} {}
+
+  int32_t serializedSize() const {
+    return bloomFilter.serializedSize();
+  }
+
+  void serialize(char* output) const {
+    return bloomFilter.serialize(output);
+  }
+
+  void mergeWith(StringView& serialized) {
+    bloomFilter.merge(serialized.data());
+  }
+
+  bool initialized() {
+    return bloomFilter.isSet();
+  }
+
+  void init(int32_t capacity) {
+    if (!bloomFilter.isSet()) {
+      bloomFilter.reset(capacity);
+    }
+  }
+
+  void insert(int64_t value) {
+    bloomFilter.insert(folly::hasher<int64_t>()(value));
+  }
+
+  BloomFilter<StlAllocator<uint64_t>> bloomFilter;
+}; // namespace
+
+class BloomFilterAggAggregate : public exec::Aggregate {
+ public:
+  explicit BloomFilterAggAggregate(const TypePtr& resultType)
+      : Aggregate(resultType) {}
+
+  int32_t accumulatorFixedWidthSize() const override {
+    return sizeof(BloomFilterAccumulator);
+  }
+
+  bool isFixedSize() const override {
+    return false;
+  }
+
+  void initializeNewGroups(
+      char** groups,
+      folly::Range<const vector_size_t*> indices) override {
+    setAllNulls(groups, indices);
+    for (auto i : indices) {
+      new (groups[i] + offset_) BloomFilterAccumulator(allocator_);
+    }
+  }
+
+  void addRawInput(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushdown*/) override {
+    decodeArguments(rows, args);
+    VELOX_USER_CHECK(
+        !decodedRaw_.mayHaveNulls(), "First argument value should not be null");
```

Comment:
> First argument value

Users may not understand what this refers to. How about,

```
First argument of bloom_filter_agg cannot be null
```

However, !decodedRaw_.mayHaveNulls() is a very strong check. It may return false even if there are no nulls.

Reply:
I'm a bit confused about ` It may return false even if there are no nulls.`
In my mind, I understand it is supposed to be `return false if there are no nulls, return false or true when there are nulls or not`, so it is determined there is no nulls.

Other codes obey this rule.
https://github.com/facebookincubator/velox/blob/main/velox/connectors/hive/HiveConnector.cpp#L633
Scalar function document:
https://facebookincubator.github.io/velox/develop/scalar-functions.html
```
bool mayHaveNulls() : constant time check on the underlying vector nullity. When it returns false, there are definitely no nulls, a true does not guarantee null existence.
```